starknet_py/net/schemas/rpc.py

```diff
@@ -510,3 +522,172 @@ class PendingTransactionsSchema(Schema):
     @post_load
     def make_dataclass(self, data, **kwargs):
         return data["pending_transactions"]
+
+
+# ------------------------------- Trace API -------------------------------
+
+
+class EventContentSchema(Schema):
+    keys = fields.List(Felt(), data_key="keys", required=True)
+    data = fields.List(Felt(), data_key="data", required=True)
+
+    @post_load
+    def make_dataclass(self, data, **kwargs):
+        return EventContent(**data)
+
+
+class FunctionInvocationSchema(Schema):
+    contract_address = Felt(data_key="contract_address", required=True)
+    entry_point_selector = Felt(data_key="entry_point_selector", required=True)
+    calldata = fields.List(Felt(), data_key="calldata", required=True)
+    caller_address = Felt(data_key="caller_address", required=True)
+    class_hash = Felt(data_key="class_hash", required=True)
+    entry_point_type = EntryPointTypeField(data_key="entry_point_type", required=True)
+    call_type = CallTypeField(data_key="call_type", required=True)
+    result = fields.List(Felt(), data_key="result", required=True)
+    # https://marshmallow.readthedocs.io/en/stable/nesting.html#nesting-a-schema-within-itself
+    calls = fields.List(
+        fields.Nested(
+            lambda: FunctionInvocationSchema()  # pylint: disable=unnecessary-lambda
+        ),
+        data_key="calls",
+        required=True,
+    )
+    events = fields.List(
+        fields.Nested(EventContentSchema()), data_key="events", required=True
+    )
+    messages = fields.List(
+        fields.Nested(L2toL1MessageSchema()), data_key="messages", required=True
+    )
+
+    @post_load
+    def make_dataclass(self, data, **kwargs) -> FunctionInvocation:
+        return FunctionInvocation(**data)
+
+
+class ExecuteInvocationSchema(Schema):
+    revert_reason = fields.String(data_key="revert_reason", load_default=None)
+    function_invocation = fields.Dict(data_key="function_invocation", load_default=None)
+
+    @pre_load
+    def alter_data(self, data, **kwargs):
+        # This method does data preprocessing before it is loaded into the fields above. If tx is not reverted, data
+        # gets loaded into 'function_invocation' data by adding a dict layer, otherwise gets loaded into "revert_reason"
+        assert isinstance(data, dict)
+        if "revert_reason" not in data:
+            data = {"function_invocation": data}
+        return data
+
+    @post_load
+    def make_dataclass_or_string(
+        self, data, **kwargs
+    ) -> Union[FunctionInvocation, dict]:
+        if "revert_reason" not in data:
+            return FunctionInvocation(**data["function_invocation"])
+        return {"revert_reason": data["revert_reason"]}
```

Comment:
Like i mentioned before, maybe it could get simpler if a dataclass with `revert_reason` is used instead.